// @stencil/router/dist/esm/es5/build/ijlk0ldf.entry.js
import { b as ActiveRouter } from './chunk-30abd2e3.js';
var RouteTitle = /** @class */ (function () {
    function RouteTitle() {
        this.titleSuffix = '';
        this.pageTitle = '';
    }
    RouteTitle.prototype.updateDocumentTitle = function () {
        document.title = "" + this.pageTitle + (this.titleSuffix || '');
    };
    RouteTitle.prototype.componentWillLoad = function () {
        this.updateDocumentTitle();
    };
    Object.defineProperty(RouteTitle, "is", {
        get: function () { return "stencil-route-title"; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RouteTitle, "properties", {
        get: function () {
            return {
                "el": {
                    "elementRef": true
                },
                "pageTitle": {
                    "type": String,
                    "attr": "page-title",
                    "watchCallbacks": ["updateDocumentTitle"]
                },
                "titleSuffix": {
                    "type": String,
                    "attr": "title-suffix"
                }
            };
        },
        enumerable: true,
        configurable: true
    });
    return RouteTitle;
}());
ActiveRouter.injectProps(RouteTitle, [
    'titleSuffix',
]);
export { RouteTitle as StencilRouteTitle };
