// @stencil/router/dist/esm/es5/build/ils9pxev.sc.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../stencilrouter.core.js';
var AsyncContent = /** @class */ (function () {
    function AsyncContent() {
        this.content = '';
    }
    AsyncContent.prototype.componentWillLoad = function () {
        if (this.documentLocation != null) {
            return this.fetchNewContent(this.documentLocation);
        }
    };
    AsyncContent.prototype.fetchNewContent = function (newDocumentLocation) {
        var _this = this;
        return fetch(newDocumentLocation)
            .then(function (response) { return response.text(); })
            .then(function (data) {
            _this.content = data;
        });
    };
    AsyncContent.prototype.render = function () {
        return (h("div", { innerHTML: this.content }));
    };
    Object.defineProperty(AsyncContent, "is", {
        get: function () { return "stencil-async-content"; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AsyncContent, "properties", {
        get: function () {
            return {
                "content": {
                    "state": true
                },
                "documentLocation": {
                    "type": String,
                    "attr": "document-location",
                    "watchCallbacks": ["fetchNewContent"]
                }
            };
        },
        enumerable: true,
        configurable: true
    });
    return AsyncContent;
}());
export { AsyncContent as StencilAsyncContent };
