// @stencil/router/dist/esm/es5/build/n9lxjggs.entry.js
import { b as ActiveRouter } from './chunk-30abd2e3.js';
var StencilRouterPrompt = /** @class */ (function () {
    function StencilRouterPrompt() {
        this.when = true;
        this.message = '';
    }
    StencilRouterPrompt.prototype.enable = function (message) {
        if (this.unblock) {
            this.unblock();
        }
        if (this.history) {
            this.unblock = this.history.block(message);
        }
    };
    StencilRouterPrompt.prototype.disable = function () {
        if (this.unblock) {
            this.unblock();
            this.unblock = undefined;
        }
    };
    StencilRouterPrompt.prototype.componentWillLoad = function () {
        if (this.when) {
            this.enable(this.message);
        }
    };
    StencilRouterPrompt.prototype.updateMessage = function (newMessage, prevMessage) {
        if (this.when) {
            if (!this.when || prevMessage !== newMessage) {
                this.enable(this.message);
            }
        }
        else {
            this.disable();
        }
    };
    StencilRouterPrompt.prototype.componentDidUnload = function () {
        this.disable();
    };
    StencilRouterPrompt.prototype.render = function () {
        return null;
    };
    Object.defineProperty(StencilRouterPrompt, "is", {
        get: function () { return "stencil-router-prompt"; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StencilRouterPrompt, "properties", {
        get: function () {
            return {
                "el": {
                    "elementRef": true
                },
                "history": {
                    "type": "Any",
                    "attr": "history"
                },
                "message": {
                    "type": String,
                    "attr": "message",
                    "watchCallbacks": ["updateMessage"]
                },
                "unblock": {
                    "state": true
                },
                "when": {
                    "type": Boolean,
                    "attr": "when",
                    "watchCallbacks": ["updateMessage"]
                }
            };
        },
        enumerable: true,
        configurable: true
    });
    return StencilRouterPrompt;
}());
ActiveRouter.injectProps(StencilRouterPrompt, [
    'history',
]);
export { StencilRouterPrompt };
