// @stencil/router/dist/esm/es5/build/zedbp4to.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../stencilrouter.core.js';
import { a as matchPath } from './chunk-fea4bd05.js';
import { a as isModifiedEvent } from './chunk-88bd89b0.js';
import { b as ActiveRouter } from './chunk-30abd2e3.js';
import './chunk-cd950d93.js';
function getUrl(url, root) {
    if (url.charAt(0) == '/' && root.charAt(root.length - 1) == '/') {
        return root.slice(0, root.length - 1) + url;
    }
    return root + url;
}
var RouteLink = /** @class */ (function () {
    function RouteLink() {
        this.unsubscribe = function () { return; };
        this.activeClass = 'link-active';
        this.exact = false;
        this.strict = true;
        this.custom = 'a';
        this.match = null;
    }
    RouteLink.prototype.componentWillLoad = function () {
        this.computeMatch();
    };
    RouteLink.prototype.computeMatch = function () {
        if (this.location) {
            this.match = matchPath(this.location.pathname, {
                path: this.urlMatch || this.url,
                exact: this.exact,
                strict: this.strict
            });
        }
    };
    RouteLink.prototype.handleClick = function (e) {
        if (isModifiedEvent(e) || !this.history || !this.url || !this.root) {
            return;
        }
        e.preventDefault();
        return this.history.push(getUrl(this.url, this.root));
    };
    RouteLink.prototype.render = function () {
        var _a;
        var anchorAttributes = {
            class: (_a = {},
                _a[this.activeClass] = this.match !== null,
                _a),
            onClick: this.handleClick.bind(this)
        };
        if (this.anchorClass) {
            anchorAttributes.class[this.anchorClass] = true;
        }
        if (this.custom === 'a') {
            anchorAttributes = Object.assign({}, anchorAttributes, { href: this.url, title: this.anchorTitle, role: this.anchorRole, tabindex: this.anchorTabIndex, 'aria-haspopup': this.ariaHaspopup, id: this.anchorId, 'aria-posinset': this.ariaPosinset, 'aria-setsize': this.ariaSetsize, 'aria-label': this.ariaLabel });
        }
        return (h(this.custom, Object.assign({}, anchorAttributes), h("slot", null)));
    };
    Object.defineProperty(RouteLink, "is", {
        get: function () { return "stencil-route-link"; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RouteLink, "properties", {
        get: function () {
            return {
                "activeClass": {
                    "type": String,
                    "attr": "active-class"
                },
                "anchorClass": {
                    "type": String,
                    "attr": "anchor-class"
                },
                "anchorId": {
                    "type": String,
                    "attr": "anchor-id"
                },
                "anchorRole": {
                    "type": String,
                    "attr": "anchor-role"
                },
                "anchorTabIndex": {
                    "type": String,
                    "attr": "anchor-tab-index"
                },
                "anchorTitle": {
                    "type": String,
                    "attr": "anchor-title"
                },
                "ariaHaspopup": {
                    "type": String,
                    "attr": "aria-haspopup"
                },
                "ariaLabel": {
                    "type": String,
                    "attr": "aria-label"
                },
                "ariaPosinset": {
                    "type": String,
                    "attr": "aria-posinset"
                },
                "ariaSetsize": {
                    "type": Number,
                    "attr": "aria-setsize"
                },
                "custom": {
                    "type": String,
                    "attr": "custom"
                },
                "el": {
                    "elementRef": true
                },
                "exact": {
                    "type": Boolean,
                    "attr": "exact"
                },
                "history": {
                    "type": "Any",
                    "attr": "history"
                },
                "location": {
                    "type": "Any",
                    "attr": "location",
                    "watchCallbacks": ["computeMatch"]
                },
                "match": {
                    "state": true
                },
                "root": {
                    "type": String,
                    "attr": "root"
                },
                "strict": {
                    "type": Boolean,
                    "attr": "strict"
                },
                "url": {
                    "type": String,
                    "attr": "url"
                },
                "urlMatch": {
                    "type": String,
                    "attr": "url-match"
                }
            };
        },
        enumerable: true,
        configurable: true
    });
    return RouteLink;
}());
ActiveRouter.injectProps(RouteLink, [
    'history',
    'location',
    'root'
]);
export { RouteLink as StencilRouteLink };
