// @stencil/router/dist/esm/es2017/build/n9lxjggs.sc.entry.js
/*! Built with http://stenciljs.com */
import { h } from '../stencilrouter.core.js';

import { b as ActiveRouter } from './chunk-30abd2e3.js';

class StencilRouterPrompt {
    constructor() {
        this.when = true;
        this.message = '';
    }
    enable(message) {
        if (this.unblock) {
            this.unblock();
        }
        if (this.history) {
            this.unblock = this.history.block(message);
        }
    }
    disable() {
        if (this.unblock) {
            this.unblock();
            this.unblock = undefined;
        }
    }
    componentWillLoad() {
        if (this.when) {
            this.enable(this.message);
        }
    }
    updateMessage(newMessage, prevMessage) {
        if (this.when) {
            if (!this.when || prevMessage !== newMessage) {
                this.enable(this.message);
            }
        }
        else {
            this.disable();
        }
    }
    componentDidUnload() {
        this.disable();
    }
    render() {
        return null;
    }
    static get is() { return "stencil-router-prompt"; }
    static get properties() { return {
        "el": {
            "elementRef": true
        },
        "history": {
            "type": "Any",
            "attr": "history"
        },
        "message": {
            "type": String,
            "attr": "message",
            "watchCallbacks": ["updateMessage"]
        },
        "unblock": {
            "state": true
        },
        "when": {
            "type": Boolean,
            "attr": "when",
            "watchCallbacks": ["updateMessage"]
        }
    }; }
}
ActiveRouter.injectProps(StencilRouterPrompt, [
    'history',
]);

export { StencilRouterPrompt };
